File: services/entity-resolution-service/src/fusion_base/domain/engine/graph_resolution.rs
use std::collections::{BTreeMap, BTreeSet};

use chrono::Utc;
use uuid::Uuid;

use crate::models::cluster::{EntityRecord, MatchEvidence, ResolvedCluster, ReviewQueueItem};

pub struct GraphResolutionResult {
    pub clusters: Vec<ResolvedCluster>,
    pub review_items: Vec<ReviewQueueItem>,
}

pub fn resolve_clusters(
    job_id: Uuid,
    records: &[EntityRecord],
    evidences: &[MatchEvidence],
    review_threshold: f32,
    auto_merge_threshold: f32,
) -> GraphResolutionResult {
    let mut union_find = UnionFind::new(records.len());
    let record_positions = records
        .iter()
        .enumerate()
        .map(|(index, record)| (record.record_id.clone(), index))
        .collect::<BTreeMap<_, _>>();

    for evidence in evidences {
        if evidence.final_score < review_threshold {
            continue;
        }

        let Some(left_index) = record_positions.get(&evidence.left_record_id) else {
            continue;
        };
        let Some(right_index) = record_positions.get(&evidence.right_record_id) else {
            continue;
        };

        union_find.union(*left_index, *right_index);
    }

    let mut grouped = BTreeMap::<usize, Vec<EntityRecord>>::new();
    for (index, record) in records.iter().enumerate() {
        grouped
            .entry(union_find.find(index))
            .or_default()
            .push(record.clone());
    }

    let mut clusters = Vec::new();
    let mut review_items = Vec::new();
    let now = Utc::now();

    for records in grouped.into_values() {
        let cluster_id = Uuid::now_v7();
        let record_ids = records
            .iter()
            .map(|record| record.record_id.clone())
            .collect::<BTreeSet<_>>();

        let cluster_evidence = evidences
            .iter()
            .filter(|evidence| {
                record_ids.contains(&evidence.left_record_id)
                    && record_ids.contains(&evidence.right_record_id)
                    && evidence.final_score >= review_threshold
            })
            .cloned()
            .collect::<Vec<_>>();

        let requires_review = cluster_evidence.iter().any(|evidence| {
            evidence.final_score >= review_threshold && evidence.final_score < auto_merge_threshold
        });

        let confidence_score = if cluster_evidence.is_empty() {
            let total_confidence = records.iter().map(|record| record.confidence).sum::<f32>();
            (total_confidence / records.len().max(1) as f32).clamp(0.0, 1.0)
        } else {
            let total_score = cluster_evidence
                .iter()
                .map(|evidence| evidence.final_score)
                .sum::<f32>();
            (total_score / cluster_evidence.len().max(1) as f32).clamp(0.0, 1.0)
        };

        let status = if cluster_evidence.is_empty() {
            "singleton"
        } else if requires_review {
            "pending_review"
        } else {
            "resolved"
        };

        let cluster_key = records
            .iter()
            .map(|record| record.record_id.clone())
            .collect::<Vec<_>>()
            .join("|");

        let cluster = ResolvedCluster {
            id: cluster_id,
            job_id,
            cluster_key,
            status: status.to_string(),
            records: records.clone(),
            evidence: cluster_evidence.clone(),
            confidence_score,
            requires_review,
            suggested_golden_record_id: None,
            created_at: now,
            updated_at: now,
        };

        if requires_review {
            let rationale = cluster_evidence
                .iter()
                .take(3)
                .map(|evidence| evidence.explanation.clone())
                .collect::<Vec<_>>();

            review_items.push(ReviewQueueItem {
                id: Uuid::now_v7(),
                cluster_id,
                status: "pending".to_string(),
                severity: if confidence_score < auto_merge_threshold {
                    "high"
                } else {
                    "medium"
                }
                .to_string(),
                recommended_action: "manual_review".to_string(),
                rationale,
                assigned_to: None,
                reviewed_by: None,
                notes: String::new(),
                created_at: now,
                updated_at: now,
            });
        }

        clusters.push(cluster);
    }

    clusters.sort_by(|left, right| right.records.len().cmp(&left.records.len()));
    GraphResolutionResult {
        clusters,
        review_items,
    }
}

struct UnionFind {
    parents: Vec<usize>,
    ranks: Vec<usize>,
}

impl UnionFind {
    fn new(size: usize) -> Self {
        Self {
            parents: (0..size).collect(),
            ranks: vec![0; size],
        }
    }

    fn find(&mut self, index: usize) -> usize {
        if self.parents[index] != index {
            let root = self.find(self.parents[index]);
            self.parents[index] = root;
        }
        self.parents[index]
    }

    fn union(&mut self, left: usize, right: usize) {
        let left_root = self.find(left);
        let right_root = self.find(right);
        if left_root == right_root {
            return;
        }

        if self.ranks[left_root] < self.ranks[right_root] {
            self.parents[left_root] = right_root;
        } else if self.ranks[left_root] > self.ranks[right_root] {
            self.parents[right_root] = left_root;
        } else {
            self.parents[right_root] = left_root;
            self.ranks[left_root] += 1;
        }
    }
}

File: openfoundry-go/services/entity-resolution-service/internal/domain/engine/graph_resolution.go
package engine

import (
	"sort"
	"strings"
	"time"

	"github.com/google/uuid"

	"github.com/openfoundry/openfoundry-go/services/entity-resolution-service/internal/models"
)

// GraphResolutionResult mirrors fusion_base::domain::engine::graph_resolution::GraphResolutionResult.
type GraphResolutionResult struct {
	Clusters    []models.ResolvedCluster
	ReviewItems []models.ReviewQueueItem
}

// ResolveClusters ports `graph_resolution::resolve_clusters` exactly:
// union-find over records linked by evidences whose final_score >=
// review_threshold, then cluster bookkeeping + review item synthesis.
func ResolveClusters(jobID uuid.UUID, records []models.EntityRecord, evidences []models.MatchEvidence, reviewThreshold, autoMergeThreshold float32) GraphResolutionResult {
	uf := newUnionFind(len(records))
	positions := make(map[string]int, len(records))
	for i, record := range records {
		positions[record.RecordID] = i
	}

	for _, evidence := range evidences {
		if evidence.FinalScore < reviewThreshold {
			continue
		}
		li, lok := positions[evidence.LeftRecordID]
		ri, rok := positions[evidence.RightRecordID]
		if !lok || !rok {
			continue
		}
		uf.union(li, ri)
	}

	groups := map[int][]models.EntityRecord{}
	for i, record := range records {
		root := uf.find(i)
		groups[root] = append(groups[root], record)
	}
	rootKeys := make([]int, 0, len(groups))
	for k := range groups {
		rootKeys = append(rootKeys, k)
	}
	sort.Ints(rootKeys)

	clusters := []models.ResolvedCluster{}
	reviewItems := []models.ReviewQueueItem{}
	now := time.Now().UTC()

	for _, root := range rootKeys {
		clusterRecords := groups[root]
		clusterID := MustNewUUIDv7()
		idSet := map[string]struct{}{}
		for _, r := range clusterRecords {
			idSet[r.RecordID] = struct{}{}
		}

		clusterEvidence := []models.MatchEvidence{}
		for _, evidence := range evidences {
			_, lok := idSet[evidence.LeftRecordID]
			_, rok := idSet[evidence.RightRecordID]
			if lok && rok && evidence.FinalScore >= reviewThreshold {
				clusterEvidence = append(clusterEvidence, evidence)
			}
		}

		requiresReview := false
		for _, evidence := range clusterEvidence {
			if evidence.FinalScore >= reviewThreshold && evidence.FinalScore < autoMergeThreshold {
				requiresReview = true
				break
			}
		}

		var confidenceScore float32
		if len(clusterEvidence) == 0 {
			total := float32(0)
			for _, r := range clusterRecords {
				total += r.Confidence
			}
			denom := len(clusterRecords)
			if denom < 1 {
				denom = 1
			}
			confidenceScore = clamp01(total / float32(denom))
		} else {
			total := float32(0)
			for _, evidence := range clusterEvidence {
				total += evidence.FinalScore
			}
			denom := len(clusterEvidence)
			if denom < 1 {
				denom = 1
			}
			confidenceScore = clamp01(total / float32(denom))
		}

		var status string
		switch {
		case len(clusterEvidence) == 0:
			status = "singleton"
		case requiresReview:
			status = "pending_review"
		default:
			status = "resolved"
		}

		recordIDs := make([]string, 0, len(clusterRecords))
		for _, r := range clusterRecords {
			recordIDs = append(recordIDs, r.RecordID)
		}
		clusterKey := strings.Join(recordIDs, "|")

		cluster := models.ResolvedCluster{
			ID:                      clusterID,
			JobID:                   jobID,
			ClusterKey:              clusterKey,
			Status:                  status,
			Records:                 append([]models.EntityRecord(nil), clusterRecords...),
			Evidence:                append([]models.MatchEvidence(nil), clusterEvidence...),
			ConfidenceScore:         confidenceScore,
			RequiresReview:          requiresReview,
			SuggestedGoldenRecordID: nil,
			CreatedAt:               now,
			UpdatedAt:               now,
		}

		if requiresReview {
			rationale := []string{}
			for i, evidence := range clusterEvidence {
				if i >= 3 {
					break
				}
				rationale = append(rationale, evidence.Explanation)
			}
			severity := "medium"
			if confidenceScore < autoMergeThreshold {
				severity = "high"
			}
			reviewItems = append(reviewItems, models.ReviewQueueItem{
				ID:                MustNewUUIDv7(),
				ClusterID:         clusterID,
				Status:            "pending",
				Severity:          severity,
				RecommendedAction: "manual_review",
				Rationale:         rationale,
				AssignedTo:        nil,
				ReviewedBy:        nil,
				Notes:             "",
				CreatedAt:         now,
				UpdatedAt:         now,
			})
		}

		clusters = append(clusters, cluster)
	}

	sort.SliceStable(clusters, func(i, j int) bool {
		return len(clusters[i].Records) > len(clusters[j].Records)
	})

	return GraphResolutionResult{Clusters: clusters, ReviewItems: reviewItems}
}

type unionFind struct {
	parents []int
	ranks   []int
}

func newUnionFind(size int) *unionFind {
	parents := make([]int, size)
	for i := range parents {
		parents[i] = i
	}
	return &unionFind{parents: parents, ranks: make([]int, size)}
}

func (u *unionFind) find(index int) int {
	if u.parents[index] != index {
		u.parents[index] = u.find(u.parents[index])
	}
	return u.parents[index]
}

func (u *unionFind) union(left, right int) {
	lr := u.find(left)
	rr := u.find(right)
	if lr == rr {
		return
	}
	switch {
	case u.ranks[lr] < u.ranks[rr]:
		u.parents[lr] = rr
	case u.ranks[lr] > u.ranks[rr]:
		u.parents[rr] = lr
	default:
		u.parents[rr] = lr
		u.ranks[lr]++
	}
}

// MustNewUUIDv7 returns a UUIDv7 (matches Rust `Uuid::now_v7()`).
// On the (extremely unlikely) failure of `uuid.NewV7`, falls back to
// UUIDv4 — same panic-free behaviour the Rust impl gives.
func MustNewUUIDv7() uuid.UUID {
	id, err := uuid.NewV7()
	if err != nil {
		return uuid.New()
	}
	return id
}
